### PR TITLE
chore: Parallelize unit tests

### LINKED ISSUE #1431: Unit tests are not parallelized
Unit tests are not parallelized, leading to slower test execution times.

### PR DESCRIPTION
**Pull request by orpheuslummis**
## Relevant issue(s)

Partially resolves #1431

## Description

Increase parallelization of unit tests

Skipped packages because they had issues when parallelized (e.g. data race)
- config
- events
- api/http
- node

Note that "nearly all" of the test execution time is not in the unit tests, but rather in the integration test suite.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

CI

### CODE REVIEW

File: client/errors_test.go
Comment (AndrewSisley) [inline client/errors_test.go:20]: nitpick: This line very much appears to be useless, as it is the only such call within the package

File: merkle/crdt/merklecrdt_test.go
Comment (AndrewSisley) [inline merkle/crdt/merklecrdt_test.go:43]: nitpick: This line very much appears to be useless, as it is the only such call within the package

**Review by AndrewSisley — APPROVED** [review]
Approving because it has a tangible positive impact.  

I really really wish that this didnt require a call per test function though, it is really ugly IMO and I am annoyed at the Go dev team for not coming up with something nicer.

Might be worth waiting on another approval before merge, as this does IMO have a negative impact on the maintainability of the code (is ugly, and requires conscious thought, especially if a test stops working because of it (do you remove parallel, or change the prod code)).

File: client/errors_test.go
Comment (orpheuslummis) [reply client/errors_test.go:20]: what about leaving it as a default norm to for the tests to be defined?

Comment (AndrewSisley) [reply client/errors_test.go:20]: I really dont think we should suggest or enforce that, it is essentially dead, untested, code.  It suggests that the test runs in parallel, but it does not, and in many cases the ability to run that test in parallel has/will not be tested. The introduction of a second Parallel call may cause the first/original test to fail.

Comment (orpheuslummis) [reply client/errors_test.go]: resolved

File: merkle/crdt/merklecrdt_test.go
Comment (orpheuslummis) [reply merkle/crdt/merklecrdt_test.go]: resolved

File: client/errors_test.go
Comment (orpheuslummis) [reply client/errors_test.go]: *trivia*: in Rust tests run in parallel by default, but can be made sequential with a synchronization mechanism or `#[serial]` attribute from crate https://docs.rs/serial_test/

Comment (AndrewSisley) [reply client/errors_test.go]: Yes, I'm a bigger fan of this way round.  Opting out of running them in parallel is more like accepting (and warning/flagging) that the production code is broken and needs fixing.

**Comment by jsimnz** [comment]
Considering the majority of our test time is in the integration, what is the decrease in test time with this change?

**Comment by orpheuslummis** [comment]
> Considering the majority of our test time is in the integration, what is the decrease in test time with this change?

probably quite negligible 

most usefulness comes from identifying potential brittle areas in the system

**Pull request closed without merging**
